modules/dataControllerModule/index.js

```diff
@@ -0,0 +1,195 @@
+/**
+ * This module validates the configuration and filters data accordingly
+ * @module modules/dataController
+ */
+import {config} from '../../src/config.js';
+import {getHook, module} from '../../src/hook.js';
+import {deepAccess, prefixLog, deepSetValue} from '../../src/utils.js';
+import {startAuction} from '../../src/prebid.js';
+
+const LOG_PRE_FIX = 'Data_Controller : ';
+const ALL = '*';
+const MODULE_NAME = 'dataController';
+const GLOBAL = 'global';
+let _dataControllerConfig;
+
+const _logger = prefixLog(LOG_PRE_FIX);
+
+/**
+ * BidderRequests hook to intiate module and reset data object
+ */
+export function filterBidData(fn, req) {
+  if (_dataControllerConfig.filterEIDwhenSDA) {
+    filterEIDs(req.adUnits, req.ortb2Fragments);
+  }
+
+  if (_dataControllerConfig.filterSDAwhenEID) {
+    filterSDA(req.adUnits, req.ortb2Fragments);
+  }
+  fn.call(this, req);
+  return req;
+}
+
+function containsConfiguredEIDS(eidSourcesMap, bidderCode) {
+  if (_dataControllerConfig.filterSDAwhenEID.includes(ALL)) {
+    return true;
+  }
+  let bidderEIDs = eidSourcesMap.get(bidderCode);
+  if (bidderEIDs == undefined) {
+    return false;
+  }
+  let containsEIDs = false;
+  _dataControllerConfig.filterSDAwhenEID.forEach(source => {
+    if (bidderEIDs.has(source)) {
+      containsEIDs = true;
+    }
+  });
+  return containsEIDs;
+}
+
+function containsConfiguredSDA(segementMap, bidderCode) {
+  if (_dataControllerConfig.filterEIDwhenSDA.includes(ALL)) {
+    return true;
+  }
+  return hasValue(segementMap.get(bidderCode)) || hasValue(segementMap.get(GLOBAL))
+}
+
+function hasValue(bidderSegement) {
+  let containsSDA = false;
+  if (bidderSegement == undefined) {
+    return false;
+  }
+  _dataControllerConfig.filterEIDwhenSDA.forEach(segment => {
+    if (bidderSegement.has(segment)) {
+      containsSDA = true;
+    }
+  });
+  return containsSDA;
+}
+
+function getSegmentConfig(ortb2Fragments) {
+  let bidderSDAMap = new Map();
```

Comment:
there's some duplication in here, which you could avoid by reusing the same logic for both global and bidder data. Here's a possible refactor to illustrate what I mean:

```
let bidderSDAMap = new Map();
function collectSegments(key, data) {
   let segmentSet = constructSegment(deepAccess(data, 'user.data') || []);
   if (segmentSet && segmentSet.size > 0) bidderSDAMap.set(key, segmentSet);
}
collectSegments(GLOBAL, ortb2Fragments.global); 
Object.entries(ortb2Fragments.bidder).forEach(([bidder, data]) => collectSegments(bidder, data));
return bidderSDAMap;
```

Reply:
Fixed